File: backend/app/schemas/mobile_inspection.py
from datetime import datetime
from typing import List

from pydantic import BaseModel, Field


class StartIn(BaseModel):
    source: str | None = Field(default="manual", pattern="^(manual|qr)$")


class AttachmentBrief(BaseModel):
    id: int
    file_name: str
    url: str
    file_size: int | None = None
    category: str | None = None


class ItemResultIn(BaseModel):
    check_item_id: int
    value: str | None = None
    is_abnormal: bool = False
    remark: str | None = None


class SaveEquipmentResultIn(BaseModel):
    result: str = Field(..., pattern="^(normal|abnormal)$")
    issue_description: str | None = None
    items: List[ItemResultIn] = Field(default_factory=list)


class CheckItemBrief(BaseModel):
    id: int
    item_code: str
    item_name: str
    input_type: str
    standard_value: str | None = None
    unit: str | None = None


class EquipmentItemValue(BaseModel):
    check_item_id: int
    item_code: str
    item_name: str
    input_type: str
    standard_value: str | None = None
    unit: str | None = None
    value: str | None = None
    is_abnormal: bool = False
    remark: str | None = None


class EquipmentBriefForInspect(BaseModel):
    id: int
    equipment_code: str
    equipment_name: str
    equipment_type: str
    equipment_type_label: str
    location: str | None = None
    result: str | None = None  # null / normal / abnormal
    completed_at: datetime | None = None
    issue_description: str | None = None
    item_count: int
    abnormal_item_count: int


class StartOut(BaseModel):
    record_id: int
    record_no: str
    status: str
    source: str
    inspection_time: datetime
    room: dict
    inspector: dict
    equipment_list: List[EquipmentBriefForInspect]
    progress: dict  # { total, completed, normal, abnormal }
    completed_count: int
    total_count: int
    remaining_count: int
    next_pending_equipment_id: int | None = None
    all_completed: bool


class EquipmentDetailOut(BaseModel):
    record_id: int
    current_record: dict | None = None
    equipment: dict
    items: List[EquipmentItemValue]
    attachments: List[AttachmentBrief]
    result: str | None = None
    issue_description: str | None = None
    completed_at: datetime | None = None
    equipment_list: List[EquipmentBriefForInspect] = Field(default_factory=list)
    progress: dict | None = None
    completed_count: int | None = None
    total_count: int | None = None
    remaining_count: int | None = None
    next_pending_equipment_id: int | None = None
    all_completed: bool | None = None


class ProgressOut(BaseModel):
    total: int
    completed: int
    normal: int
    abnormal: int


class SaveEquipmentResultOut(BaseModel):
    record_id: int
    completed_count: int
    total_count: int
    remaining_count: int
    next_equipment_id: int | None = None
    next_pending_equipment_id: int | None = None
    all_completed: bool
    progress: ProgressOut


class SubmitOut(BaseModel):
    record_id: int
    record_no: str
    status: str
    has_issue: int
    submitted_at: datetime
    summary: dict  # equipment counts

File: frontend/src/api/mobileInspection.ts
import request, { api } from './request'
import { getToken } from '@/utils/auth'

export interface InspectorBrief {
  id: number
  name: string
  role: string
}

export interface InspectionRoomBrief {
  id: number
  code: string
  name: string
  area?: string | null
}

export interface InspectionEquipmentBrief {
  id: number
  equipment_code: string
  equipment_name: string
  equipment_type: string
  equipment_type_label: string
  location?: string | null
  result: 'normal' | 'abnormal' | null
  completed_at?: string | null
  issue_description?: string | null
  item_count: number
  abnormal_item_count: number
}

export interface InspectionProgress {
  total: number
  completed: number
  normal: number
  abnormal: number
}

export interface StartInspectionData {
  record_id: number
  record_no: string
  status: string
  source: 'manual' | 'qr'
  inspection_time: string
  room: InspectionRoomBrief
  inspector: InspectorBrief
  equipment_list: InspectionEquipmentBrief[]
  progress: InspectionProgress
  completed_count: number
  total_count: number
  remaining_count: number
  next_pending_equipment_id: number | null
  all_completed: boolean
}

/** One enabled room in the mobile picker, with whether the current account already inspected it today. */
export interface InspectableRoom {
  id: number
  code: string
  name: string
  area?: string | null
  owner_id?: number | null
  owner_name?: string | null
  status: number
  equipment_count: number
  inspected_today: boolean
}

export const apiInspectableRooms = () =>
  api.get<InspectableRoom[]>('/api/mobile/inspection/rooms')

export const apiStartRoomInspection = (roomCode: string, source: 'manual' | 'qr' = 'manual') =>
  api.post<StartInspectionData>(`/api/mobile/inspection/rooms/${encodeURIComponent(roomCode)}/start`, { source })

export interface AttachmentBrief {
  id: number
  file_name: string
  url: string
  file_size?: number
  category?: string | null
}

export interface EquipmentItemValue {
  check_item_id: number
  item_code: string
  item_name: string
  input_type: 'boolean' | 'number' | 'text' | 'photo'
  standard_value?: string | null
  unit?: string | null
  value: string | null
  is_abnormal: boolean
  remark: string | null
}

export interface EquipmentInspectionDetail {
  record_id: number
  current_record?: {
    id: number
    record_no: string
    status: string
    room_id: number
  }
  equipment: {
    id: number
    equipment_code: string
    equipment_name: string
    equipment_type: string
    equipment_type_label: string
    location?: string | null
  }
  items: EquipmentItemValue[]
  attachments: AttachmentBrief[]
  result: 'normal' | 'abnormal' | null
  issue_description: string | null
  completed_at: string | null
  equipment_list?: InspectionEquipmentBrief[]
  progress?: InspectionProgress
  completed_count?: number
  total_count?: number
  remaining_count?: number
  next_pending_equipment_id?: number | null
  all_completed?: boolean
}

export const apiGetEquipmentInspection = (recordId: number, equipmentId: number) =>
  api.get<EquipmentInspectionDetail>(`/api/mobile/inspection/records/${recordId}/equipment/${equipmentId}`)

export interface ItemResultPayload {
  check_item_id: number
  value: string | null
  is_abnormal: boolean
  remark?: string | null
}

export interface SaveEquipmentResultPayload {
  result: 'normal' | 'abnormal'
  issue_description?: string | null
  items: ItemResultPayload[]
}

export interface SaveEquipmentResultResp {
  record_id: number
  completed_count: number
  total_count: number
  remaining_count: number
  next_equipment_id: number | null
  next_pending_equipment_id: number | null
  all_completed: boolean
  progress: InspectionProgress
}

export const apiSaveEquipmentResult = (
  recordId: number,
  equipmentId: number,
  payload: SaveEquipmentResultPayload,
) =>
  api.put<SaveEquipmentResultResp>(
    `/api/mobile/inspection/records/${recordId}/equipment/${equipmentId}`,
    payload,
  )

/** Uploads an image via multipart/form-data. Returns the persisted AttachmentBrief. */
export async function apiUploadEquipmentImage(
  recordId: number,
  equipmentId: number,
  file: File,
): Promise<AttachmentBrief> {
  const fd = new FormData()
  fd.append('file', file)
  const res = await request.post(
    `/api/mobile/inspection/records/${recordId}/equipment/${equipmentId}/attachments`,
    fd,
    {
      headers: {
        Authorization: `Bearer ${getToken()}`,
        'Content-Type': 'multipart/form-data',
      },
    },
  )
  return (res as unknown as { data: AttachmentBrief }).data
}

export const apiDeleteAttachment = (attachmentId: number) =>
  api.delete<{ id: number; deleted: boolean; file_deleted: boolean }>(
    `/api/mobile/inspection/attachments/${attachmentId}`,
  )

export interface SubmitInspectionData {
  record_id: number
  record_no: string
  status: string
  has_issue: number
  submitted_at: string
  summary: {
    equipment_total: number
    normal_equipment: number
    abnormal_equipment: number
  }
  remark?: string | null
}

export const apiSubmitInspection = (recordId: number, remark?: string) =>
  api.post<SubmitInspectionData>(
    `/api/mobile/inspection/records/${recordId}/submit`,
    remark ? { remark } : {},
  )
